### PR TITLE
Feature/query cxx types

### PR DESCRIPTION
**Pull request by rmclaren**
## Description

Adds support for a more flexible type system. BUFR field META data is inspected and the appropriate type is selected (could be float, int, uint, int64, uint64, or string). In addition the user has the ability to override the type in the bufr variables definition. 

### Issue(s) addressed

#965 
#977 
#976
#978

## Acceptance Criteria (Definition of Done)

NETCDF file data fields will be represented in more appropriate types. The user will have the ability to override the types if necessary.

## Dependencies

#1006

### CODE REVIEW

**Review by CoryMartin-NOAA — COMMENTED** [review]
Looks good to me, approval contingent on tests/builds passing though

**Comment by rmclaren** [comment]
I removed support for unsigned types in order to ensure compatibility of output with the fortran clients.

**Comment by emilyhcliu** [comment]
@rmclaren Are we still accept PRs for feature/query_cxx_types?

**Comment by rmclaren** [comment]
@emilyhcliu In general I'd like to declare this branch as done so that we can code review and merge it as soon as possible as it contains a really important feature. Since you just want to update a YAML file and some meta-data (and there hasn't been a lot of code reviews) I think its OK to merge what you've got.

**Comment by BenjaminRuston** [comment]
@rmclaren and @CoryMartin-NOAA ,,, is this the lead blocking issue now.

Get this in and more dominoes will fall?

File: src/bufr/BufrParser/Query/Constants.h [hunk 1]
```diff
@@ -9,6 +9,6 @@
 namespace Ingester {
 namespace bufr {
     /// \brief The missing data value for all BUFR data.
-    const float MissingValue = 10e10;
+    const double MissingValue = 10.0e10;
```
Comment (gthompsnJCSDA) [inline src/bufr/BufrParser/Query/Constants.h:12]: Not sure it matters, but would this make more sense as 10.0d10 since it is declared double precision?
I also admit some confusion because I thought BUFR had a single-precision missing value since a double missing would require more space to store than a single-precision.  Is this actually a NCEP bufr-lib thing not actually something inherent to BUFR storage itself?

File: src/bufr/BufrParser/Query/DataProvider.cpp [hunk 2]
```diff
@@ -98,5 +101,60 @@ namespace bufr {
     {
         delete_table_data_f();
     }
+
+    TypeInfo DataProvider::getTypeInfo(FortranIdx idx) const
+    {
+        static const unsigned int UNIT_STR_LEN = 24;
+        static const unsigned int DESC_STR_LEN = 55;
```
Comment (gthompsnJCSDA) [inline src/bufr/BufrParser/Query/DataProvider.cpp:108]: Is 55 a hard-wired limit?  Kinda arbitrary.  This may be entirely unimportant, but I often like to declare string lengths to remain multiples of 8 for memory reasons; so could this be 64?  No idea what downstream effects would occur.

File: src/bufr/README.md [hunk 3]
```diff
@@ -67,6 +67,7 @@ Defines how to read data from the input BUFR file. Its sections are as follows:
               hoursFromUtc: 0  # Optional
           satellite_id:
             query: "*/SAID"
+            type: int64
```
Comment (gthompsnJCSDA) [inline src/bufr/README.md:70]: Is this done solely for testing purposes because the list of satellite identifiers is just an integer and I would find it wildly crazy if the number of values in this table exceeds the limits of single-precision integers (roughly a billion).

File: src/bufr/BufrParser/Query/Query.cpp [hunk 4]
```diff
@@ -97,16 +96,15 @@ namespace bufr {
         maskCache_.insert({dataProvider_.getSubset(), masks});
     }
 
-    __details::Target Query::findTarget(const std::string &targetName,
-                                        const std::string &query) const {
+    std::shared_ptr<Target> Query::findTarget(const std::string &targetName,
```
Comment (nicholasesposito) [inline src/bufr/BufrParser/Query/Query.cpp:99]: Looks like accesses to structures now go through a pointer. Why the change to a pointer?

**Review by gthompsnJCSDA — APPROVED** [review]
I am approving the PR.  I had some simple inconsequential comments to make, but nothing that I think is ultra important.  I do not have sufficient C++ experience to bring a strong review to the code changes themselves.  This is a big PR and lots of changes. But, I believe the team at NCEP has done strong work to make this work.  Hats off to Ron for the tremendous amount of work here.  Forcing some data with scaling (or offsets) that require a type change from INT to FLOAT is important and this is a welcome addition.

File: src/bufr/DataObject.h [hunk 5]
```diff
@@ -372,6 +399,131 @@ namespace Ingester
             typename std::enable_if<!std::is_arithmetic<T>::value, U>::type* = nullptr) const
         {
             throw std::runtime_error("The stored value was is not a number");
+            return 0.0f;
+        }
+
+        /// \brief Set the data associated with this data object (numeric DataObject).
```
Comment (nicholasesposito) [inline src/bufr/DataObject.h:405]: What's the purpose of this new block of code?

**Review by nicholasesposito — APPROVED** [review]
The main things I saw in my review:

Lots of new use of pointers.
Addition of type information and forced overrides.
Added Target.h for gathering info/metadata when we execute a query.
Changes to DataObject to sets data for some of the objects.

Just had one question that I posted. Otherwise I think it looks good. Good job.

File: src/bufr/DataObject.h [hunk 6]
```diff
@@ -372,6 +399,131 @@ namespace Ingester
...
+                        data_[i] = static_cast<T>(static_cast<double>(data_[i]) * val);
+                    }
+                }
+            }
+            else
+            {
+                std::ostringstream str;
+                str << "Multiplying integer field \"" << fieldName_ << "\" with a non-integer is ";
+                str << "illegal. Please convert it to a float or double.";
+                throw std::runtime_error(str.str());
```
Comment (srherbener) [inline src/bufr/DataObject.h:482]: I realize that I do not fully understand the problems or context you are trying to solve with the data typing improvements in this PR. I apologize for having this limited understanding, but to me these new functions you added (setData, multiplyBy, offsetBy, etc.) seem overly complex since they seem to be trying to replace the compile time type matching/checking you get with templates with runtime type checking.

For example the multiplyBy routine here is set to use a factor that is of type double. Why not use a factor of type T instead. This removes the need to do the checking and type conversions (static_casting). Ie, the compiler will create a version with float, another with int, etc. according to what the rest of the code instantiates. In this scheme the non-numeric case (string) could be handled with a template specialization of DataObject that simply throws exceptions if you try to call multiplyBy with strings.

File: src/bufr/DataObject.h (hunk 5 above)
Comment (BenjaminRuston) [reply src/bufr/DataObject.h:405]: @rmclaren should we move forward and strip out if not needed

or even better this is something needed and useful, right?

**Review by srherbener — COMMENTED** [review]
Thanks for all the hard work to provide this update. From the comments of others, this is a very useful update which is great.
